### PR TITLE
OAK-9434 - MongoDB indexing: implement parallel chunk download

### PR DESCRIPTION
**Pull request by averma21**
**Download task creation/splitting and result collection explanation** - 

NEW_TASK_QUEUE=MultithreadedTraverseWithSortStrategy#taskQueue
FINISHED_QUEUE=TraverseAndSortTask#completedTasks

1. Have a BlockingQueue of tasks (NEW_TASK_QUEUE) that need to be executed. Each of this task has been assigned a _modified range.
2. Main thread (MultithreadedTraverseWithSortStrategy) adds an initial breakup of tasks (of type TraverseAndSortTask) to this queue.
3. Another monitoring thread (MultithreadedTraverseWithSortStrategy#TaskRunner) is consuming from this NEW_TASK_QUEUE and submitting those tasks to executor service.
4. The executor service is a fixed thread pool executor so that number of threads remains unchanged overtime.
5. There is another queue for tasks that have finished execution - FINISHED_QUEUE. Each task adds itself or its ID to this queue when it finishes execution. This is ConcurrentLinkedQueue because running tasks monitor this queue for finished tasks but don't need to block if no finished task is there.
6.  Every running task monitors FINISHED_QUEUE as part of its execution cycle. If a running task has potential of splitting - identified by what range of _modified values is left to be examined and downloaded - and if there is a task in the FINISHED_QUEUE, the running task tries to dequeue the finished task.
7. Only one of the running tasks would be able to dequeue a finished task. The running task which succeeds will reduce its upper limit of _modified range and submit a new task with the _modified range after that (till its original upper limit).
8. The monitoring thread is reading from NEW_TASK_QUEUE (blocking queue), so it needs an indication about when to stop looking for new tasks from this queue. This is done by putting a poison pill entry in this queue. The main thread does it as explained below.
9. We use a phaser for coordination between main thread, the running tasks and the monitoring thread. This phaser has two phases -
	1. Waiting for new tasks (from task splits)
	2. Waiting for result collection
10. Each task registers itself (in phase 1 of phaser) inside its constructor itself. This ensures there is no created task which is not registered to a phaser. Also when the task finishes (either normally or due to exception), it arrives and deregisters (since it is not needed in phase 2 of phaser).
11. Main thread after adding initial tasks to NEW_TASK_QUEUE, starts waiting for phase 1 to advance. Phase 1 would advance only if all the registered (and hence created) tasks have arrived. When all tasks arrive, it means we won't have any more new tasks from task splitting. 
12. So, once phase 1 advances, main thread adds the poison pill to NEW_TASK_QUEUE, reading which the monitoring thread advances to result collection. Before starting result collection, it registers itself to phaser (now in phase 2). 
13. Main thread after adding poison pill, starts waiting for phase 2 to advance. This advance happens when monitoring thread aggregates results from all tasks and adds it to data structure which main thread could access.
14. After the advance of phase 2, the main thread returns the result - the list of sorted files which are then merged by the next step of indexing process.

### CODE REVIEW

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseWithSortStrategy.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import java.lang.management.MemoryUsage;
 import java.nio.charset.Charset;
 import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseWithSortStrategy.java]: unused import

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentTraverser.java [hunk 2]
```diff
@@ -30,6 +30,8 @@
 import org.apache.jackrabbit.oak.plugins.document.NodeDocument;
 import org.apache.jackrabbit.oak.plugins.document.cache.NodeDocumentCache;
 import org.apache.jackrabbit.oak.plugins.document.util.CloseableIterable;
+import org.bson.BsonDocument;
+import org.bson.BsonString;
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentTraverser.java]: unused import

**Review by fabriziofortino — APPROVED** [review]
LGTM. Just to minors.

**Review by thomasmueller — APPROVED** [review]
This PR looks good!

Just that I don't forget: We have seen cases where the connection to MongoDB was temporarily lost (due to upscale / downscale); this also needs to be taken care of at some point.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/IndexOptions.java [hunk 3]
```diff
@@ -102,6 +103,8 @@ public IndexOptions(OptionParser parser){
         //Set of options which define action
         actionOpts = ImmutableSet.of(stats, definitions, consistencyCheck, dumpIndex, reindex, importIndex);
         operationNames = collectionOperationNames(actionOpts);
+        existingDataDumpDirOpt = parser.accepts("existing-data-dump-dir", "Directory containing mongo document dumps from previous incomplete run")
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/IndexOptions.java]: I would avoid the term "mongo document" and use "document store". For multiple reasons: the dumps don't contain any MongoDB documents actually, but Oak data. Then, at some point we might also support RDBMS (I know currently it needs MongoDocumentStore - but that's an implementation detail)

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java [hunk 4]
```diff
@@ -78,15 +80,62 @@ protected void setProviders() throws IOException {
...
 
+    private static class MongoNodeStateEntryTraverserFactory implements NodeStateEntryTraverserFactory {
+
+        private static final AtomicInteger traverserInstanceCounter = new AtomicInteger(0);
+        final RevisionVector rootRevision;
+        final DocumentNodeStore documentNodeStore;
+        final MongoDocumentStore documentStore;
+        final Logger traversalLogger;
+        final CompositeIndexer indexer;
+        final Closer closer;
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java]: I think all of those fields can also be private, right (didn't verify)?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java [hunk 5]
```diff
@@ -78,15 +80,62 @@ protected void setProviders() throws IOException {
         this.indexerProviders = createProviders();
     }
 
+    private static class MongoNodeStateEntryTraverserFactory implements NodeStateEntryTraverserFactory {
+
+        private static final AtomicInteger traverserInstanceCounter = new AtomicInteger(0);
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:89]: It is a bit weird to use a static (global) variable... I wonder if it's needed? If this is to help debugging, then that's good enough. But if we want to keep it static, then I wonder if it should be all caps.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ */
+package org.apache.jackrabbit.oak.index.indexer.document;
+
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+
+public class NodeStateEntryBuilder {
+
+    private final NodeState nodeState;
+    private final String path;
+    private long memUsage = 0;
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java]: "= 0" is not needed as 0 is the default value.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    private long lastModified = 0;
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java]: Same here

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        this.path = path;
+    }
+
+    public NodeStateEntryBuilder withMemUsage(long memUsage) {
+        this.memUsage = memUsage;
+        return this;
+    }
+
+    public NodeStateEntryBuilder withLastModified(Long lastModified) {
+        if (lastModified != null) {
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java]: Using "null" as a special case is a bit unusual. Could it be avoided? It is only called in one place, so maybe change the logic there? So that "long" can be used instead of "Long".

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.index.indexer.document;
+
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+
+public class NodeStateEntryBuilder {
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryBuilder.java]: Just an idea: instead of a separate builder class, you could probably also make the builder an inner class of NodeStateEntry - that way, the constructor of NodeStateEntry could be private (or at least package private)

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java [hunk 10]
```diff
@@ -78,15 +80,62 @@ protected void setProviders() throws IOException {
...
+        final RevisionVector rootRevision;
+        final DocumentNodeStore documentNodeStore;
+        final MongoDocumentStore documentStore;
+        final Logger traversalLogger;
+        final CompositeIndexer indexer;
+        final Closer closer;
+
+        public MongoNodeStateEntryTraverserFactory(RevisionVector rootRevision, DocumentNodeStore documentNodeStore,
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java]: I would probably use the name "DocumentStoreNodeStateEntryTraverserFactory"... I see currently we rely on MongoDocumentStore, but I think in the future this could be changed.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    private final long lastModifiedLowerBound;
+    private final long lastModifiedUpperBound;
+    private final boolean isUpperBoundExclusive;
+
+    public LastModifiedRange(long lastModifiedLowerBound, long lastModifiedUpperBound) {
+        this(lastModifiedLowerBound, lastModifiedUpperBound, true);
+    }
+
+    public LastModifiedRange(long lastModifiedLowerBound, long lastModifiedUpperBound, boolean isUpperBoundExclusive) {
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java]: I would simplify the logic here and hardcode that isUpperBoundExclusive is always true. You could rename lastModifiedLowerBound => lastModifiedFrom, and lastModifiedUpperBound => lastModifiedTo. This is the logic used in the JDK, e.g. java.util.Arrays.copyOfRange and many other places. Because having the isUpperBoundExclusive flag would mean that we need to add a lot more unit tests for this class - if we implement both behaviours, we need to test both as well.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                : lastModifiedUpperBound < range.getLastModifiedLowerBound();
+        rangeOnLeft = isUpperBoundExclusive ? range.getLastModifiedUpperBound() <= lastModifiedLowerBound :
+                range.getLastModifiedUpperBound() < lastModifiedLowerBound;
+        return !(rangeOnLeft || rangeOnRight);
+    }
+
+    @Override
+    public String toString() {
+        return "LastModifiedRange{" +
+                "lastModifiedLowerBound=" + lastModifiedLowerBound +
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java]: Personally, I would use the mathematical interval notation - https://en.wikipedia.org/wiki/Interval_(mathematics)

  LastModifiedRange [10, 20)

I think it is intuitive... It means 10 inclusive, and 20 exclusive. But maybe "LastModifiedRange(10, 20)" is better... not sure.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java [hunk 13]
```diff
@@ -76,57 +84,82 @@
...
     }
 
-    private File sortStoreFile() throws IOException {
-        log.info("Proceeding to perform merge of {} sorted files", sortedFiles.size());
-        Stopwatch w = Stopwatch.createStarted();
-        File sortedFile = new File(storeDir, getSortedStoreFileName(compressionEnabled));
-        try(BufferedWriter writer = createWriter(sortedFile, compressionEnabled)) {
-            Function<String, NodeStateHolder> func1 = (line) -> line == null ? null : new SimpleNodeStateHolder(line);
-            Function<NodeStateHolder, String> func2 = holder -> holder == null ? null : holder.getLine();
-            ExternalSort.mergeSortedFiles(sortedFiles,
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java]: I don't currently see where we sort... we do need to sort, right?

**Review by thomasmueller — CHANGES_REQUESTED** [review]
The disadvantage (if I understand correctly) is that we basically lose the current logic - were we don't use the modified index. I think this is a bit sad... What about if we continue to support the existing logic for some time, as "legacy" logic? That way, if we are unsure if a slowdown is due to using the modified index, we can still switch to the old logic.

Only for some time (a few months). Later we can remove the old code. (Otherwise we will have to maintain both for a long time).

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java (hunk 13 above)
Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java]: Yes, individual files are sorted in the method - `sortAndSaveBatch` in this class itself.
The sorted merge is done in `MultithreadedTraversalWithSortStrategy#sortStoreFile`

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/IndexOptions.java (hunk 3 above)
Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/IndexOptions.java]: done

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java (hunk 5 above)
Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:89]: Yes i actually added this for better debugging information in logs.
Regarding all caps, i think it is generally used for constants (static and final). While these two modifiers are used here, the value here is not constant (only the reference is). So i didn't capitalize the name :)

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java (hunk 4 above)
Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java]: yes. done.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java (hunk 10 above)
Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java]: I think naming this class generically at present would conflict with the fact that it used MongoDocumentStore currently. IMO generic naming should be done when we extract the specific things out and put the common stuff in generic/abstract/parent class and the specific stuff in concrete/child classes. 
This class is declared private, so that also limits the impact currently. I have also made the constructor private node.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        long latest = cursor.next().getLong(NodeDocument.MODIFIED_IN_SECS);
+        return simpleSplit(oldest, latest, parts);
+    }
+
+    public static List<Long> simpleSplit(long start, long end, int parts) {
+        long stepSize = (end - start)/parts;
+        List<Long> steps = new ArrayList<>();
+        StringBuilder splitPoints = new StringBuilder();
+        for (long i = start; i <= end; i+=stepSize) {
```
Comment (tihom88) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java:76]: Are we loosing some documents here?
lets say start =0, end =11, parts =3
stepsize= 10-0/4 = 3.666.... (float) = 3 (long)
stepslist= {0,3,6,9} ---> we don't have 10 and 11 in this list. 
Are we handling it at some later stage?

Comment (averma21) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java:76]: Yes, that is handled later. We use the range [9, LONG.MAX_VALUE)
See MultithreadedTraverseWithSortStrategy#createInitialTasks
The logic is explained in javadocs of MultithreadedTraverseWithSortStrategy as well.

Comment (tihom88) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java:76]: So we are not using checkpoint info (time) in these tasks. Are we downloading complete content and then indexing till checkpoint?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java (hunk 5 above)
Comment (thomasmueller) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:89]: Makes sense... I don't have very strong opinion if this should be traverserInstanceCounter or TRAVERSER_INSTANCE_COUNTER. I'm fine with using whatever you prefer.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java [hunk 15]
```diff
@@ -78,15 +80,62 @@ protected void setProviders() throws IOException {
...
+            this.traversalLogger = traversalLogger;
+            this.indexer = indexer;
+            this.closer = closer;
+        }
+
+        @Override
+        public NodeStateEntryTraverser create(LastModifiedRange lastModifiedRange) {
+            IndexingProgressReporter progressReporterPerTask =
+                    new IndexingProgressReporter(IndexUpdateCallback.NOOP, NodeTraversalCallback.NOOP);
+            String entryTraverserID = "NSET" + traverserInstanceCounter.incrementAndGet();
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java]: Could we use a constant for "NSET"?
What does "NSET" stand for? That could be a javadoc.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MemoryManager.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+         */
+        SELF_MANAGED
+    }
+
+    Type getType();
+
+    /**
+     * Register a client with this memory manager. All registered clients are informed when available memory level is low.
+     * If memory level is already low, client is not registered.
+     * NOTE - this method should only be used with {@link Type#JMX_BASED} instance types, otherwise the behaviour is undefined.
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MemoryManager.java]: I would just write that "otherwise, it has no effect". Or maybe reverse it and say, "for SELF_MANAGED, this method has no effect".

It is easier to understand I think. I don't see a reason for undefined behavior...

Comment (thomasmueller) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MemoryManager.java]: I think it is very good to move memory management to a different class BTW.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+
+import static org.apache.jackrabbit.oak.commons.IOUtils.humanReadableByteCount;
+import static org.apache.jackrabbit.oak.index.indexer.document.flatfile.FlatFileStoreUtils.sizeOf;
+import static org.apache.jackrabbit.oak.index.indexer.document.flatfile.MultithreadedTraverseWithSortStrategy.DirectoryHelper;
+
+/**
+ * A callable representing a task for traversing/downloading the nodes states from the node store and creating sorted files
+ * based on the downloaded data.
+ */
+class TraverseAndSortTask implements Callable<List<File>>, MemoryManagerClient {
```
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java:53]: So this was originally oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/StoreAndSortStrategy.java

It is a bit hard to review code changes when files are renamed... Would it be possible to rename files back? And then move them in a separate commit?

**Review by thomasmueller — CHANGES_REQUESTED** [review]

**Review by thomasmueller — APPROVED** [review]

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.index.indexer.document;
+
+public class LastModifiedRange {
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/LastModifiedRange.java:21]: this is okay, but you could have used https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Range.html

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryTraverser.java [hunk 19]
```diff
@@ -127,7 +155,7 @@ private boolean includeDoc(NodeDocument doc) {
 
     private CloseableIterable<NodeDocument> findAllDocuments() {
         return new MongoDocumentTraverser(documentStore)
-                .getAllDocuments(Collection.NODES, id -> includeId(id));
+                .getAllDocuments(Collection.NODES, lastModifiedRange, id -> includeId(id));
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/NodeStateEntryTraverser.java]: nitpick:  you can use method reference here

```suggestion
                .getAllDocuments(Collection.NODES, lastModifiedRange, this::includeId);
```

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java [hunk 20]
```diff
@@ -40,22 +42,42 @@
...
     private Set<String> preferredPathElements = Collections.emptySet();
     private BlobStore blobStore;
     private PathElementComparator comparator;
     private NodeStateEntryWriter entryWriter;
+    private NodeStateEntryTraverserFactory nodeStateEntryTraverserFactory;
     private long entryCount = 0;
 
-    private boolean useZip = Boolean.valueOf(System.getProperty(OAK_INDEXER_USE_ZIP, "true"));
-    private boolean useTraverseWithSort = Boolean.valueOf(System.getProperty(OAK_INDEXER_TRAVERSE_WITH_SORT, "true"));
+    private final boolean useZip = Boolean.valueOf(System.getProperty(OAK_INDEXER_USE_ZIP, "true"));
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java]: replace with `parseBoolean` to avoid unnecessary boxing:

```suggestion
    private final boolean useZip = Boolean.parseBoolean(System.getProperty(OAK_INDEXER_USE_ZIP, "true"));
```

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTaskFactory.java [hunk 21]
```diff
@@ -0,0 +1,9 @@
+package org.apache.jackrabbit.oak.index.indexer.document.flatfile;
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTaskFactory.java]: 1- Missing license here.
2- This interface seems to be not used.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+            steps.add(i);
+            splitPoints.append(" ").append(i);
+        }
+        log.info("Split points of _modified values {}", splitPoints.toString());
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/DocumentStoreSplitter.java:80]: no need to call `toString` here

```suggestion
        log.info("Split points of _modified values {}", splitPoints);
```

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreTest.java [hunk 23]
```diff
@@ -20,34 +20,75 @@
 package org.apache.jackrabbit.oak.index.indexer.document.flatfile;
 
 import java.io.File;
+import java.lang.reflect.Field;
```
Comment (fabriziofortino) [inline oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreTest.java]: unused, remove

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreTest.java [hunk 24]
```diff
@@ -20,34 +20,75 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
```
Comment (fabriziofortino) [inline oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreTest.java]: unused, remove

File: oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/ElasticIndexerTest.java [hunk 25]
```diff
@@ -61,9 +61,9 @@ public void nodeIndexed_WithIncludedPaths() throws Exception {
 
         NodeState testNode = EMPTY_NODE.builder().setProperty("foo", "bar").getNodeState();
 
-        assertTrue(indexer.index(new NodeStateEntry(testNode, "/content/x")));
-        assertFalse(indexer.index(new NodeStateEntry(testNode, "/x")));
-        assertFalse(indexer.index(new NodeStateEntry(testNode, "/")));
+        assertTrue(indexer.index(new NodeStateEntryBuilder(testNode, "/content/x").build()));
```
Comment (fabriziofortino) [inline oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/ElasticIndexerTest.java]: Does this compile for you? I get

```
[64,38] cannot find symbol
[ERROR]   symbol:   class NodeStateEntryBuilder
[ERROR]   location: class org.apache.jackrabbit.oak.index.indexer.document.ElasticIndexerTest
```

**Review by fabriziofortino — APPROVED** [review]
Good work in general. Just a few minors.

**Comment by averma21** [comment]
Thanks for reviews. I am closing this PR and will open another one from this branch to apache project directly.
